linera-base/src/identifiers.rs

```diff
@@ -276,6 +318,60 @@ impl<A> Debug for ApplicationId<A> {
             .finish()
     }
 }
+#[derive(Serialize, Deserialize)]
+#[serde(rename = "UserApplicationId")]
+struct SerializableApplicationId {
+    pub bytecode_id: BytecodeId,
+    pub creation: MessageId,
+}
+
+impl<A> Serialize for ApplicationId<A> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        if serializer.is_human_readable() {
+            let bytes = bcs::to_bytes(&SerializableApplicationId {
+                bytecode_id: self.bytecode_id.forget_abi(),
+                creation: self.creation,
+            })
+            .map_err(serde::ser::Error::custom)?;
+            serializer.serialize_str(&hex::encode(bytes))
+        } else {
+            let mut state = serializer.serialize_struct("UserApplicationId", 2)?;
+            state.serialize_field("bytecode_id", &self.bytecode_id)?;
+            state.serialize_field("creation", &self.creation)?;
+            state.end()
```

Comment:
You could use the implementation derived for `SerializableApplicationId` here, too.
To avoid cloning, you could make the fields `Cow`s.